src-rocksdb/RocksDB.hs

```diff
@@ -0,0 +1,313 @@
+optionsIncreaseParallelism (Options ptr) v =
+    rocksdb_options_increase_parallelism ptr (intToCInt v)
+
+optionsSetMaxOpenFiles :: Options -> Int -> IO ()
+optionsSetMaxOpenFiles (Options ptr) v =
+    rocksdb_options_set_max_open_files ptr (intToCInt v)
+
+optionsOptimizeLevelStyleCompaction :: Options -> Word64 -> IO ()
+optionsOptimizeLevelStyleCompaction (Options ptr) v =
+    rocksdb_options_optimize_level_style_compaction ptr v
+
+optionsSetCompression :: Options -> Int -> IO ()
+optionsSetCompression (Options ptr) v =
+    rocksdb_options_set_compression ptr (intToCInt v)
+
+-------------------------------------------------------------------------------
+-- read options
+-------------------------------------------------------------------------------
+
+newtype ReadOptions = ReadOptions (Ptr READOPTIONS)
+
+withReadOptions :: (ReadOptions -> IO r) -> IO r
+withReadOptions kont = bracket
+    rocksdb_readoptions_create
+    rocksdb_readoptions_destroy
+    (\ptr -> kont (ReadOptions ptr))
+
+-------------------------------------------------------------------------------
+-- write options
+-------------------------------------------------------------------------------
+
+newtype WriteOptions = WriteOptions (Ptr WRITEOPTIONS)
+
+withWriteOptions :: (WriteOptions -> IO r) -> IO r
+withWriteOptions kont = bracket
+    rocksdb_writeoptions_create
+    rocksdb_writeoptions_destroy
+    (\ptr -> kont (WriteOptions ptr))
+
+writeOptionsDisableWAL :: WriteOptions -> Bool {- ^ Disable -} -> IO ()
+writeOptionsDisableWAL (WriteOptions opts) disable =
+    rocksdb_writeoptions_disable_WAL opts (if disable then 1 else 0)
+
+-------------------------------------------------------------------------------
+-- db operations
+-------------------------------------------------------------------------------
+
+data RocksDB = RocksDB (Ptr DB) ErrPtr
+
+withRocksDB :: Options -> FilePath -> (RocksDB -> IO r) -> IO r
+withRocksDB (Options opt) path kont =
+    withCString path $ \path' ->
+    withErrPtr $ \errptr ->
+    bracket (rocksdb_open' path' errptr) rocksdb_close $ \ptr ->
+    kont (RocksDB ptr errptr)
+  where
+    rocksdb_open' path' errptr = do
+        ptr <- rocksdb_open opt path' errptr
+        assertErrPtr "rocksdb_open" errptr
+        return ptr
+
+put :: RocksDB -> WriteOptions -> ByteString -> ByteString -> IO ()
+put (RocksDB dbPtr errPtr) (WriteOptions opts) key val =
+    unsafeUseAsCStringLen key $ \(kp, kl) ->
+    unsafeUseAsCStringLen val $ \(vp, vl) -> do
+        rocksdb_put dbPtr opts kp (intToCSize kl) vp (intToCSize vl) errPtr
+        assertErrPtr "rocksdb_put" errPtr
+
+get :: RocksDB -> ReadOptions -> ByteString -> IO (Maybe ByteString)
+get (RocksDB dbPtr errPtr) (ReadOptions opts) key =
+    unsafeUseAsCStringLen key $ \(kp, kl) ->
+    alloca $ \vlPtr -> do
+        vp <- rocksdb_get dbPtr opts kp (intToCSize kl) vlPtr errPtr
+        assertErrPtr "rocksdb_get" errPtr -- TODO: may leak
```

Comment:
May leak what?

Reply:
the result, i.e. memory allocated by `rocksdb_get`.

Again, the docs are not clear, whether it can be possible that some value is allocated, and that `errPtr` is set too.